recursion/compiler/src/asm/instruction.rs

```diff
@@ -0,0 +1,288 @@
+use alloc::collections::BTreeMap;
+use alloc::format;
+use alloc::string::String;
+use core::fmt;
+use sp1_recursion_core::cpu::Instruction;
+use sp1_recursion_core::runtime::Opcode;
+
+use crate::builder::Builder;
+use crate::ir::Felt;
+use crate::util::canonical_i32_to_field;
+use p3_field::PrimeField32;
+
+#[derive(Debug, Clone)]
+pub enum AsmInstruction<F> {
+    /// Load work (src, dst) : load a value from the address stored at dest(fp) into src(fp).
+    LW(i32, i32),
+    /// Store word (src, dst) : store a value from src(fp) into the address stored at dest(fp).
+    SW(i32, i32),
+    // Get immediate (dst, value) : load a value into the dest(fp).
+    IMM(i32, F),
+    /// Add, dst = lhs + rhs.
+    ADD(i32, i32, i32),
+    /// Add immediate, dst = lhs + rhs.
+    ADDI(i32, i32, F),
+    /// Subtract, dst = lhs - rhs.
+    SUB(i32, i32, i32),
+    /// Subtract immediate, dst = lhs - rhs.
+    SUBI(i32, i32, F),
+    /// Subtract value from immediate, dst = rhs - lhs.
+    SUBIN(i32, i32, F),
+    /// Multiply, dst = lhs * rhs.
+    MUL(i32, i32, i32),
+    /// Multiply immediate.
+    MULI(i32, i32, F),
+    /// Divide, dst = lhs / rhs.
+    DIV(i32, i32, i32),
+    /// Divide immediate, dst = lhs / rhs.
+    DIVI(i32, i32, F),
+    /// Divide value from immediate, dst = rhs / lhs.
+    DIVIN(i32, i32, F),
```

Comment:
It's a bit weird here that b/c get reordered from the VM-side of things, maybe better to just do DIVIN(i32, F, i32)